### ISSUE TITLE
Domain definitions via class terms in locally defined functions cause errors

### ISSUE THREAD

**Issue by McEarl** [Naproche/Naproche-SAD#18]
Defining functions locally in a proof generates an error if the domain of such a function is defined via a class term. Example:

```
Lemma. Every set is a set.
Proof.
Define f[x] = x for x in {u | u = u}.
qed.
```

When trying to check this with Naproche on the command line, I get the following error message:

```
SAD.Core.ProofTask.separation: misformed argument
CallStack (from HasCallStack):
  error, called at src/SAD/Core/ProofTask.hs:85:16 in Naproche-SAD-0.1.0.0-6t0GnBNTLK8IwgewEAZs9n:SAD.Core.ProofTask
```

Since I've inserted some comments in my ProofTask.hs, in the original version of Naproche the error should be called at src/SAD/Core/ProofTask.hs:56:16, i.e. it's `separation` which generates the error. I'm a little bit confused, because it doesn't seem that the replacement tag which is generated by a class term definition is recognized by the functions in ProofTask.hs. Could this cause the error?

**Comment by McEarl** [comment]
Well, I could fix the error myself. It's the function `SAD.Core.ProofTask.domain` which causes the problem: It handles the cases "Define f[x] = ... for x in X" and "Define f[x] = ... for x in {u in X | ...}" (where X is a predefined set in both cases), but not the case "Define f[x] = ... for x in {u | ...}". By simply adding the following line to the definition of `SAD.Core.ProofTask.domain` the latter case is handled as well.

```
domain (Tag Domain (All x (Iff _ (Tag Replacement f)))) = Tag DomainTask $ replacement (Decl.name x) f
```

But there's still the problem that such local function definitions (and also local set definitions) generate proof tasks which state that every element in a class term is contained in some set. But this often causes verification failures since Naproche doesn't regarde class terms as sets.

**Comment by anfelor** [comment]
Yes, our story about the distinction between sets and classes has become rather muddled. The old ForTheL tried to pretend that it used sets instead of classes and thus disallowed `{u | u = u}`. In comparison `{u in X | ...}` is of course a set by separation. You should talk to @PeterKoepke if this should be merged, he has been working on replacing sets by classes for quite a while. In any case, you can create a pull request with your comments to [anfelor/Naproche-SAD](https://github.com/anfelor/Naproche-SAD/); I am not sure it is possible to have too much documentation :)
